neureka/hal/neureka_task.c

```diff
@@ -0,0 +1,201 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "neureka_task.h"
+#include "neureka_task_defs.h"
+#include "pulp_nnx_util.h"
+
+inline uint32_t neureka_get_tile_padding(uint32_t padding, uint32_t i_height,
+                                         uint32_t i_width, uint32_t n_height,
+                                         uint32_t n_width) {
+  uint32_t tile_padding = padding;
+  if (i_height > 0) {
+    tile_padding &= ~(0xf << 28);
+  }
+  if (i_width < n_width - 1) {
+    tile_padding &= ~(0xf << 24);
+  }
+  if (i_height < n_height - 1) {
+    tile_padding &= ~(0xf << 20);
+  }
+  if (i_width > 0) {
+    tile_padding &= ~(0xf << 16);
+  }
+  return tile_padding;
+}
+
+void neureka_task_init(neureka_task_t *task, const uint8_t kernel_shape,
+                       const uint8_t depthwise, const uint8_t input_bits,
+                       const uint8_t output_bits, const uint8_t weights_bits,
+                       const neureka_weight_offset_mode_e weights_offset_mode,
+                       const uint32_t weights_offset_factor,
+                       neureka_quant_t quant, neureka_norm_t norm,
+                       const uint8_t flag_input_signed) {
+  *task = (neureka_task_t){.outbytes = output_bits / 8,
+                           .qw = weights_bits,
+                           .output_channel_throughput =
+                               depthwise ? NEUREKA_INPUT_CHANNEL_THROUGHPUT_3x3
+                                         : NEUREKA_OUTPUT_CHANNEL_THROUGHPUT,
+                           .input_channel_throughput =
+                               kernel_shape == 3
+                                   ? NEUREKA_INPUT_CHANNEL_THROUGHPUT_3x3
+                                   : NEUREKA_INPUT_CHANNEL_THROUGHPUT_1x1,
+                           .kernel_shape = kernel_shape,
+                           .depthwise = depthwise,
+                           .data = {0}};
+
+  const int flag_mode = kernel_shape == 1 ? NEUREKA_FLAG_MODE_1x1
+                        : depthwise == 1  ? NEUREKA_FLAG_MODE_3x3_DW
+                                          : NEUREKA_FLAG_MODE_3x3;
+
+  task->data.cfg.conf0 |=
+      flag_input_signed << NEUREKA_SHIFT_FLAG_INPUT_SIGNED |
+      NEUREKA_FLAG_NORM_QUANT | quant.function | quant.mode |
+      (quant.shift_amount << 16) |
+      quant.flag_rounding << NEUREKA_SHIFT_ROUNDING | norm.mode |
+      norm.flag_bias << NEUREKA_SHIFT_FLAG_NORM_BIAS |
+      norm.flag_shift << NEUREKA_SHIFT_FLAG_NORM_SHIFT | NEUREKA_FLAG_USE_TCDM |
+      weights_offset_mode | flag_mode | (weights_bits - 1);
```

Comment:
I'd put a link to the register map of N-EUREKA here. While this looks correct, it seems hard to verify without this information.